Refactor Ship Scheduling

Noordfrees opened the pull request:
Fixes #503 
Fixes #3544 
Fixes #3680 
Fixes #3683 
Fixes #3832 

---

Reimplements the planning-ahead shipping algorithm. All decision are made by a new class `ShippingSchedule`, which keeps a detailed list of which ship will when arrive at which port and how many items for which destinations it will pick up there, and uses this information to distribute ships to ports. This is done in six steps:
- Scan all ships. Refresh the prediction when they will arrive at the next port. Most of the time, a simple estimate is enough. Now and then (every 20s), we calculate the exact time though to account for delays resulting e.g. from ships stopping to let another ship pass.
- Scan all ports. Make lists of waiting items. Figure out when the items will be picked up. Also cancel orders where we provided more capacity than is actually needed (which can happen when a transfer is cancelled when the item is still in the portdock), and cancel expedition ships in spe whose expeditions were cancelled.
- Go through the list of ships that had orders cancelled, and check whether we might even skip some of their destinations altogether.
- First of all, check the waiting unserviced expeditions. If a ship is heading for such a port and will not pick up anything there, and has no plans beyond, make it an expedition ship there. Then go through all ports with still unserviced expeditions ready, and make a list of all idle or trivial ships. (A ship is called trivial if it is not planning to pick up any wares or service an expedition.) Assign every idle/trivial ship the closest unserviced expedition, until we run out of idle ships or all expeditions are serviced.
- Go through the lists of start-end pairs where we need more capacity. Assign each pair a priority based on the sum of the transfer priorities of the individual wares and workers.
	 - For each pair, check it a ship is coming that will visit the destination shortly afterwards and still has capacity for more items. If so, we'll assign the extra capacity to this ship, but only if the ship will go straight from here to there – and the time from now to this ship's arrival here is not too high. If the latter condition is not met, make a note of this ship.
	 - If we didn't assign the entire required capacity yet, look for idle ships and assign one or more of them (preferably the closest ones) the task of transporting those items.
	 - Still capacity left? Also accept the ships we noted in step 1.
	 - And if that still isn't enough, check if there are other ports within a low radius of the start and end ports, and also accept ships that have a destination in the start group *directly followed by* a destination in the end group, and has free capacity between these destinations, and tell such a ship to additionally visit the start and end port between its two existing targets.
- Make a list of all ships that are idle, and distribute them more or less evenly among ports: For each port, count how many ships are heading there or already located close by. Distribute the idle ships among the ports with the fewest ships: Send each ship to one of these ports (preferably a close by one).

Noordfrees committed:
Intermediate commit

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Continued working

Noordfrees committed:
Continued working

Noordfrees committed:
Saveloading support

Noordfrees committed:
Continue working

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Completed basic implementation

Noordfrees committed:
Fixed all compile errors

actions-user committed:
5 files were automatically formatted.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Fixed a swarm of bugs (more swarm to follow…)

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
More bugfixes

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Savegame compat & some bugs

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Fixing the test suite

A now invalid testcase was deleted, and another one adapted for the new behaviour

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Another small bugfix

Noordfrees committed:
Fix handling of cancelled expeditions

Noordfrees committed:
Fixing the testsuite

Noordfrees committed:
Another test suite fix

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Fixing a cornercase in port_removed()

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Fixed an error in pass 5.4

Noordfrees committed:
Fix an error on port removal

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Try to fix illegal A-B-A plans

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

klaus-halfmann commented:
going to clone and then testplay this (Europe wit a lot of Ports :-)

klaus-halfmann commented:
OK, I got a crashPlaying Europe, starting In italy (after perhpas 3 hours)


NOCOM: 
ShippingSchedule::ship_arrived(Ingeld at 123417)
Assertion failed: (ship.get_nritems() == 0), function ship_arrived, file ../src/economy/shipping_schedule.cc, line 148.
Abort trap: 6

Do we have some Bug wehre I can attach a savegame? I will try to reproduce this now

Noordfrees commented:
You can attach files right here in the PR. Just drag-and-drop the zipped .wgf file into the comment field.
Can you also attach the full log output (if you still have it)?

klaus-halfmann commented:
Here your are ... mmh, was to big youll find it here https://www.magentacloud.de/share/tu4ayusx.k

klaus-halfmann commented:
Still uploading first one. But here is another one. Will Upload to the smae location, later (~2 hour)

klaus-halfmann commented:
OK, both Crashes / Replays Uploaded, please tell me If I can start again with the save files, or must start from the Beginning.

Noordfrees commented:
The replay desyncs after 2:53:05 for me. Which commit exactly did you use for playing?

klaus-halfmann commented:
git show HEAD
commit 9f12dc387e016e146f2d4feea0f4ef68f1ec4ded (HEAD -> shipping-tweaks-b21, origin/shipping-tweaks-b21)
Merge: 7d277d878d cabc125c6d
Author: Benedikt Straub <benedikt-straub@web.de>
Date:   Thu Apr 23 22:37:20 2020 +0200
    Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

klaus-halfmann commented:
Anything I can/should do now?

Noordfrees committed:
Fixed a potential desync causer

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees commented:
Okay good news is that I found out what the desync was most likely caused by and pushed a fix. Bad news is that this means I can't use the replay to investigate the assert. So if you could try again to cause the assert with the commit I just pushed it would be great.
Oh and please play without AI opponents as having multiple fleets may make it harder to investigate the problem

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

klaus-halfmann commented:
Got another one, only logs attached, as the Replay is _so_ long, I will attach the latest logs only
[Crash.txt](https://github.com/widelands/widelands/files/4533334/Crash.txt) 
Will try to reproduce this bases on the lates autoseav tomorrow.

Noordfrees commented:
Could you please try to keep and attach the complete log? There's a **lot** of debug output but I really need all of it to trace issues properly.

klaus-halfmann commented:
Sorry the rest of the Terminal oputput, got lost ...

Noordfrees commented:
You could try to use `./widelands > output.log` to make sure all the output will be preserved. It's just that the complete log contains so much information that I don't even need to be able to reproduce an issue myself :)

frankystone commented:
Some things i found (no crashes so far):

1. If an expedition found a portspace and and all wares are unloaded from the ship, the ship immediately moves forward to another port. Nevertheless the string in the ship statistics window (opened by pressing 'e') shows: "[Shipname] founding a colony". The string persists as long the window didn't get reopened.
2. All ships never idle. They always navigate around, mostly empty (not transporting any wares) or transporting only one ware. For a player it is not possible to check if the amount of ships are enough for his economy. Visually it looks like all ships are needed, but this isn't true.

In the attached savegame, open the ship statitstics (press 'e') and found a colony. after that watch the ship navigating around.
[test shipping_tweaks.zip](https://github.com/widelands/widelands/files/4533936/test.shipping_tweaks.zip)

frankystone commented:
Empty ships, not idling
![empty_ships](https://user-images.githubusercontent.com/6730556/80290429-2e72f380-8745-11ea-9d1a-cc123c02d69b.jpg)

Noordfrees commented:
Re 1: Good spot, will fix this
Re 2: In the forum people complained that ships tend to hover around their last port when they have nothing to do. This lead to far-away ports having to wait for ages when they need a ship. My solution is to distribute all idle ships ± evenly among ports. When wares are sent from A to B but never from B to A, some ships will be sent empty away from B to keep them evenly distributed, which causes extra "busyness" of other ships as a side effect. (This happens as soon as a ship is sent to B rather than only when it arrives there). I guess there is no perfect solution, but IMHO this is better than many ships clustering around few ports.

Noordfrees committed:
Fix seafaring stats not updated and make schedule more desync-robust

actions-user committed:
2 files were automatically formatted.

klaus-halfmann commented:
OK reproduction was fast this time. Did nothing directy ship relate during till the crash hit me
[2020-04-26T07.11.01.zip](https://github.com/widelands/widelands/files/4534682/2020-04-26T07.11.01.zip) Was on the status just before b67c773066bc6efa40d7377378a70395d0722b9. Will pull and try again

klaus-halfmann commented:
I now see a warnings:
../src/economy/shipping_schedule.h:38:1: warning: struct 'MapObjectLoader' was previously declared as a class; this is valid, but may result in linker errors under the Microsoft C++ ABI 
struct MapObjectLoader;
../src/economy/shippingitem.h:32:7: note: previous use is here
class MapObjectLoader;
    AND
./src/economy/shipping_schedule.cc:783:4: warning: extra ';' after member function definition

klaus-halfmann commented:
An here is another one, I now Did play around more with a lot of Buildings
/widelands > ~/Downloads/widelands20200426_0734.log
Assertion failed: (dock2->dock != ppp.end), function update, file ../src/economy/shipping_schedule.cc, line 1123.
Abort trap: 6

[2020-04-26T07.36.54.zip](https://github.com/widelands/widelands/files/4534699/2020-04-26T07.36.54.zip)

frankystone commented:
Now i get also some crashes. E.g. loading the attached savegame and dismantling the port causes a crash.
[test shipping_tweaks.zip](https://github.com/widelands/widelands/files/4534936/test.shipping_tweaks.zip)

Noordfrees commented:
These three crashed are enough homework to keep me busy for a while :)
Fortunately kaputtnik's crash is identical to Klaus's second crash: Both are caused by a ship having the same destination multiple times in its plan ("illegal A-B-A plan") and I already have an idea where the cause is hidden.
Klaus's first crash is quite the opposite, here a ship has too few destinations so that it is unable to deliver all wares to their destinations. This one will be a bit more tricky to investigate.

Noordfrees committed:
Rewrote some buggy code and added even more log output

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees commented:
Both problems are most likely fixed now. I also added a lot more debug output, apologies for the log file size. 
These changes do not fix existing buggy savegames so be sure to start playing a fresh game or you'll experience "phantom crashes".

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Making travis happy

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

frankystone commented:
Again a crash. Don't know why but the version shows: 

This is Widelands Version r24618[7fe347d@HEAD] (Debug)

Unfortunately i have no save game :-(

Attached all available log output:
[shipping_tweaks.txt](https://github.com/widelands/widelands/files/4541707/shipping_tweaks.txt)

Noordfrees committed:
Do not allocate wares to expedition ships in spe. Fixed scoring. More debug info.

Noordfrees commented:
The crash was a simple oversight, fixed now. 
I also noticed in the log that ship scoring is broken, should also be fixed now. Added even more debug output just in case.

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

klaus-halfmann commented:
Ready for more testing, or do you still have homework?

Noordfrees commented:
All crashes reported here so far are fixed, and even travis is happy now, so ready for more testing :)

Noordfrees committed:
Found and fixed another crash

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

frankystone commented:
In the attached save game the ship Gallus will arrive at the main port, unloading 4 wares and loading one ware (a shovel). Now it is navigating to a destination where this shovel isn't needed. Instead the shovel is needed at coords 211/127, which is in the opposite direction.
[wl_autosave_00.wgf.zip](https://github.com/widelands/widelands/files/4547791/wl_autosave_00.wgf.zip)

Log after loading this savegame: 
[log.txt](https://github.com/widelands/widelands/files/4547801/log.txt)

I can provide a savegame which is 5 Minutes earlier, if you need it.

Noordfrees commented:
The schedule says
```
· Gallus: carrying 4 items (capacity 30)
  – 4 items to 2327
  SCHEDULE: 2 stations
          · in 100400 ms at 2327
            load there: cargo for 1 destinations
            – 1 items to 2655
          · in 210600 ms at 2655
            load there: cargo for 0 destinations
```
So it's carrying 4 items to port nr.2327 and will there pick up 1 item (the shovel) and take it to port no.2655. Watching the savegame I see that this is done as planned and the shovel is turned into a forester there who then reenters the portdock to be shipped to his actual destination. Looks like issue #3250 to me and is unrelated to this branch. 
(edit: never confuse a tool with the worker it will become – they are not related to each other in any way ;) )

frankystone commented:
Ahhh... messed up shovel and worker (forrester). Sorry for not being that patient to watch further.

frankystone commented:
Got another crash:

~~~~
widelands: /home/kaputtnik/Quellcode/widelands-repo/widelands/src/economy/shipping_schedule.cc:102: void Widelands::ShippingSchedule::ship_arrived(Widelands::Game&, Widelands::Ship&, Widelands::PortDock&): Zusicherung »plan_size == 1« nicht erfüllt.
Abgebrochen (Speicherabzug geschrieben)
~~~~

All available log: 
[shipping_tweaks_log.txt](https://github.com/widelands/widelands/files/4550051/shipping_tweaks_log.txt)

frankystone commented:
This savegame crashes as soon the ship called 'Venus' arrives at the port.
[test_shipping_tweaks_2.wgf.zip](https://github.com/widelands/widelands/files/4550153/test_shipping_tweaks_2.wgf.zip)

Full log: 
[shipping_log.txt](https://github.com/widelands/widelands/files/4550156/shipping_log.txt)

Noordfrees commented:
Those look like both the same crash… according to the log, the *Venus* is supposed to load an expedition at this port – and then to go on transporting wares which is not allowed, will investigate…

Noordfrees committed:
Fixed expedition detection in 5.4

Noordfrees commented:
Okay should be fixed. This does not fix the existing savegame though.

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Fixed a compiler warning

Noordfrees committed:
Improved first pass recalculation performance and fixed a b20 savegame compatibility issue

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Niektory commented:
Took f67c8e441330b87be4d7e0018637694b5f1cf525 for a spin yesterday. [Replay and log](https://github.com/widelands/widelands/files/4556064/shipping_test_04_29.zip)

Some weird behavior I observed:

2:18:40 an expedition is fully prepared at the home port, but no ship starts it until 2:44:07. Several empty ships visit the port in that time, some of them multiple times. I think expeditions shouldn't wait so long and should take priority over transporting wares.

1:55:30 Norderoogsand loads some wares at the home port, including a brick; it unloads them on the island 9, except the brick, which is taken back to the home port.

2:39:25 Oland loads some wares at the home port, and takes them on a long tour, which includes visiting the home port 2 more times. On one visit I think it doesn't even load or unload anything. It finally goes to island 12 to unload them at 2:56:10.

Newly built ships have no names in the seafaring statistics but this is probably not related to this branch.

> Re 2: In the forum people complained that ships tend to hover around their last port when they have nothing to do. This lead to far-away ports having to wait for ages when they need a ship. My solution is to distribute all idle ships ± evenly among ports. When wares are sent from A to B but never from B to A, some ships will be sent empty away from B to keep them evenly distributed, which causes extra "busyness" of other ships as a side effect. (This happens as soon as a ship is sent to B rather than only when it arrives there). I guess there is no perfect solution, but IMHO this is better than many ships clustering around few ports.

This is not a bad idea, but would it be possible for the ships that are being redistributed in this way to have a status other than "shipping"?

Niektory commented:
Can't load a f67c8e4 savegame in 2f26956, is it expected?
```
Game data error
map objects: [../src/economy/ship_fleet.cc:803] loading ship fleet: [../src/economy/shipping_schedule.cc:1596] loading shipping schedule: Stream ended unexpectedly (1 bytes read, 4 expected)
```

klaus-halfmann commented:
Played Europe (Spain) for about 5 hours, but had to stop as game was getting to slow,
will recompile as release and try more hours, once laggy again I will do "evil" things like sinking ships
or destroying harbours.

frankystone commented:
> 2:18:40 an expedition is fully prepared at the home port, but no ship starts it until 2:44:07. Several empty ships visit the port in that time, some of them multiple times. I think expeditions shouldn't wait so long and should take priority over transporting wares.

I have observed similar: A ship transports a builder to a port where the builder was the last missing item to fulfill an expedition. The ship unloads the builder, waits a moment, and then navigate to another port. The ready expedition then has to wait for another ship.

> Newly built ships have no names in the seafaring statistics but this is probably not related to this branch.

Yes i observed this also. The ship get a name after it was built, but the seafaring statistics get only updated if the ship arrives at a port.

> This is not a bad idea, but would it be possible for the ships that are being redistributed in this way to have a status other than "shipping"?

+1 Should be a new bug and targeted to build 22, imho.

Nevertheless i think this is definitely an improvement compared to build20 :)

Noordfrees commented:
- The latest commit broke savegame compatibility to older versions of this branch.
- In the same commit, I fixed a performance problem in games with many ships.
- The seafaring stats only update themselves on demand, IMHO it should be made to think() instead and update itself. But this (as well as the shipping stats text) is for b22.

> Took f67c8e4 for a spin yesterday. Replay and log

`Oland: carrying 121 items (capacity 30)` ooo-kay… That should not be possible. Did you use a release or debug build? (edit: I see release in the log. Debug builds would be better because I have added masses of asserts to fail early.)
Anyway I'll investigate
Edit: Most likely fixed now

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Need to update ScoredShips of other PPPs when loading on a PPP.

I'm curious what travis will think of this…

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Fixed a regression

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Fixed cornercases on port destruction

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

klaus-halfmann commented:
OK here is the next [Crash.txt.zip](https://github.com/widelands/widelands/files/4564145/Crash.txt.zip), will now try to reproduce this from last savegame

klaus-halfmann commented:
and with the latest changes I get
[77/101] Building CXX object src/economy/CMakeFiles/economy.dir/shipping_schedule.cc.o
../src/economy/shipping_schedule.cc:848:4: warning: extra ';' after member function definition [-Wextra-semi]
                };
                 ^
../src/economy/shipping_schedule.cc:914:36: warning: lambda capture 'this' is not used [-Wunused-lambda-capture]
        auto get_free_capacity_between = [this, &game](

Noordfrees committed:
Fixed a crash

Noordfrees commented:
Easy one, fixed crash and warnings :)

Noordfrees committed:
typo

klaus-halfmann commented:
Now I got the complets logs, will compile again, next
[2020-05-01T17.38.27.zip](https://github.com/widelands/widelands/files/4565048/2020-05-01T17.38.27.zip)

klaus-halfmann commented:
I saved this file yeterday [Turkey.wgf.zip](https://github.com/widelands/widelands/files/4567127/Turkey.wgf.zip)
 On starting the Mutiplayer Game (Human on Slot 4) it immeditaley crashes with 
 2: expedition max duration = 12600 (210 minutes), map area root: 512
Assertion failed: (&worker == working_positions_[main_worker_].worker), function get_building_work, file ../src/logic/map_objects/tribes/productionsite.cc, line 856.
Abort trap: 6

On master this is not compatible, so?

Noordfrees commented:
This is unrelated to this branch and identical to #3538 . Will have a look at it anyway.

frankystone commented:
Does this need more testing?

Noordfrees commented:
Only if you want to, it has been tested pretty extensively already so it's less likely there are critical bugs left. There can never be *too much* of testing though ;)

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

frankystone commented:
Looks like it is not tested enough ;)

~~~~
widelands: /home/kaputtnik/Quellcode/widelands-repo/widelands/src/economy/shipping_schedule.cc:460: Widelands::Duration Widelands::ShippingSchedule::update(Widelands::Game&): Zusicherung »plan.second.size() <= nr_ports« nicht erfüllt.
Abgebrochen (Speicherabzug geschrieben)
~~~~

Savegame and log attached. The savegame is from the seafaring tutorial, close before the first ship was build. I can reproduce this crash by just waiting until the first ship is ready and start transporting wares between the harbors. The crash happens after the ship navigates 4 or 5 times.

[Seefahrt (Wasserwege).wgf.zip](https://github.com/widelands/widelands/files/4595093/Seefahrt.Wasserwege.wgf.zip)

Noordfrees commented:
Ouch! This map has a rather unusual design that needs special handling, thanks for finding this…

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Fix another assert

Noordfrees committed:
Merge branch 'shipping-tweaks-b21' of github.com:Noordfrees/widelands into shipping-tweaks-b21

Noordfrees commented:
Fixed

frankystone commented:
> Ouch! This map has a rather unusual design that needs special handling, thanks for finding this…

Sounds not so good. We may have other maps with other unusual designs?

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees commented:
By "unusual" I mean: Only 2 ports which are located *barely* within a certain distance threshold of each other. My algorithm uses special treatment for close-by ports, e.g. if A+B and C+D are close by and a ship is going from A to D and a ware for C is waiting at B then the ship will be sent A-B-C-D. This caused illegal A-B-A plans to chosen in a corner case, and this corner case is most likely to surface on maps like tut03. (edit: to clarify, this was a real bug that was just more likely to surface on some maps than on others.)
Other than that the algorithm does not care about the port layout, only the distances between pairs of port are considered for decision-making.

frankystone commented:
Thanks for the explanation :-)

frankystone commented:
And another one :-(

~~~~
widelands: /home/kaputtnik/Quellcode/widelands-repo/widelands/src/economy/shipping_schedule.cc:1267: Widelands::Duration Widelands::ShippingSchedule::update(Widelands::Game&): Zusicherung »index_of_start > index_of_end« nicht erfüllt.
Abgebrochen (Speicherabzug geschrieben)
~~~~

Just played the last submitted savegame (tutorial) until the end.

The attached savegame segfaults after loading. Just wait. 

Edit: It segfaults close you have hit the ok button of the last message

[Seefahrt (Eisenerzmine).wgf.zip](https://github.com/widelands/widelands/files/4595743/Seefahrt.Eisenerzmine.wgf.zip)

Noordfrees committed:
Deleted an invalid assert

Noordfrees commented:
Easy one :) this assert was simply not true in some corner-cases which can safely be skipped.

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

frankystone commented:
Now i think this really tested enough :-)

Sorry that i can't do a code review....

Noordfrees commented:
Thanks for your testing :)

Did you also try out corner-cases like e.g.?
- Sinking ships
- Destroying some ports
- Destroying *all* ports (while there are still ships with wares left) and then rebuilding them

frankystone commented:
> * Sinking ships

Yes, but i tested this some revisions prior to the current one.

> * Destroying some ports

Yes, just did it.

> * Destroying _all_ ports (while there are still ships with wares left) and then rebuilding them

Not tested. I  am a bit tired of testing ...

Noordfrees commented:
> > * Sinking ships
> 
> Yes, but i tested this some revisions prior to the current one.

The last revs can't have affected the behaviour there so that's good

> > * Destroying some ports
> 
> Yes, just did it.

Good :)

> > * Destroying _all_ ports (while there are still ships with wares left) and then rebuilding them
> 
> Not tested. I am a bit tired of testing ...

No problem, You already tested a lot, I lost count of the number of asserts you found :)
I have tested this special case a few times without problem, and as this should happen very rarely in games it does not matter if it is not tested as rock-solidly as the normal-case behaviour…

klaus-halfmann reviewed (changes requested):
First part of Review, will to in chunks as of the amount of code.
I dont see real flaws in the code. Just improve the docs so that some
nove programmer has a chnace to get into it.

klaus-halfmann reviewed (changes requested):
2 nd part of review, only a few remarks, main code still ahead

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Addressed review

Noordfrees committed:
Added more comments

klaus-halfmann reviewed (changes requested):
More comments, some fucntions are to large and I lost track what they are about

Noordfrees committed:
Addressed review

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

Noordfrees committed:
Removed NOCOMs

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

klaus-halfmann reviewed (changes requested):
That main function is a monster, can we slice it into digestable pieces?

Noordfrees committed:
Addressed review

actions-user committed:
'src/economy/shipping_schedule.cc' was automatically formatted.

klaus-halfmann reviewed (changes requested):
Wow this exchausted me and I was only able to scan upto the end of shipping_schedule.cc.

I will do more , but to do it thoroughly we will need more time.

Some general thoughts:
* Better use explicit structs instead of std:pair as this will make the code less readadble.
* Try to make local structs / objects more "intelligent" to reduce complexity of the code
   by delegating to them.

(I wonder what the theory of logistics could tell us)

The code has a lot of assert() s. I _think_ we could avoid a lot of them. but not yet.

We can have this code, but we must rework it after release 21.
Can we setup a testcase to measure the improvements reached by this code?

@Noordfrees thanks for you coding efforts.

Noordfrees committed:
Pulled out some code

actions-user committed:
2 files were automatically formatted.

Noordfrees commented:
> The code has a lot of assert() s. I think we could avoid a lot of them. but not yet.

I'm rather in favour of having masses of asserts to ensure that we will notice as early as possible if something is fishy. If we allow code that is not working 100% as it should to continue for a while then we may eventually get a crash in a totally unrelated code part.
If we have asserts everywhere then we can be fairly certain that the bug is located in the same code part as the assert.

> We can have this code, but we must rework it after release 21.

Agreed. For now the main goal is to have a code that works well with as few bugs as possible and that will not crash. After the release there is certainly still room for improvement both in behaviour and in code style.

> Can we setup a testcase to measure the improvements reached by this code?

Our testsuite is currently using fairly simple designs, there is hardly any room for improvements there. A testcase with a large setup with many ports and long distances might be pretty hard to create…

frankystone commented:
> Can we setup a testcase to measure the improvements reached by this code?

During test playing i found that this is definitely an improvement over the current algorithm. 

Beside the fact that ships do not idle anymore, or only for a short time. This results in assuming that ships are always needed, visually, but in fact they just navigate empty between the ports. The Button for showing only idle ships in the ships statistics window is pretty useless now... For build 21 this can be as it is, but in my opinion this should be considered for build 22.

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

tppq commented:
> During test playing i found that this is definitely an improvement over the current algorithm.

I played a round of Dust in the Wind, and got the feeling that shipping was less of a pain than what it was a long time ago.

.. but the AI was too powerful and the second "mystery box" crushed me. For once I was not happy to see the AI improving.

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

Noordfrees commented:
@klaus-halfmann what's the state here?

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

klaus-halfmann commented:
Sorry I have no time or nerves at all. Got a Major incident last Weekend and nother one today.
Widelands  fo fun, which I currently dont have at all...

stonerl commented:
It seems as if there are still some problems. In the included replay/save game the yellow player had a problem were for at least an hour the ships didn't transport wares and workers. The game hung at 3:42 or so, and we had to reload from a save game and then ships started to work somewhat properly, again.

Also, there were 31 builders in one harbor on an island that didn't get picked up to start constructing. So the only way to get new builders was to destroy the harbor and loose all goods and workers. This was at 5 hour mark or so.

I'm not sure if the problem is the shipping algorithm or the combination of shipping and ferries. I just noticed the waterways are preferred over land routes. The  problem is, if there are no ferries on the waterway, wares don't get transported because they are marked for transportation over the waterways.

Another problem are islands that were only reachable via ships. There are some maps now, where waterways interfere with the old behavior. The maps Calvisson and Fellowships are an example. There are islands on these maps that were only reachable by ships. Now one can build a waterway but has to discover the islands harbor by ship.

That deeply confuses some players, but also algorithms for pathfinding and seafaring are not able to handle such conditions properly it seems.

[Replay_with_shipping_probs.zip](https://github.com/widelands/widelands/files/4707775/Replay_with_shipping_probs.zip)

Noordfrees commented:
Regarding ferries: They are treated exactly like roads by the pathfinder for wares. If you have just one waterway connecting two shores, then you have created a serious bottleneck, but this route is so much better than the alternatives that the pathfinder will prefer it of course. The sf tutorial also explains at large that in such cases it may be more efficient not to use a waterway; that one can not use WWs to claim otherwise unreachable territory, and that WWs should not be built before you have ferries for them. Doing any of these things is simply bad playing style, not much different from building one forester for ten woodcutters and then wondering at the wood shortage IMHO.

Regarding the transport issues – sorry but your post is too imprecise. Which player's port at which coords at which time has 31 builders? At which time was the one hour without transportation starting? (Watching replays is really annoying in debug builds at 3 FPS so I prefer to fast-forward to the time when something is actually going wrong…)

Niektory commented:
> WWs should not be built before you have ferries for them.

A bit off-topic, but I find this a little annoying. If you build waterways before ferries, wares get routed through them immediately and get stuck. If you build ferries before waterways, they sink if you get distracted and wait too long. I think it would be more convenient if waterways without ferries were not considered by the pathfinder.

Noordfrees commented:
Then we might have the situation that two places in the same ware economy are not reachable from one another. Only connecting two ware economies when the waterway between them has a ferry, rather than immediately when it's built, would be tricky to code and very likely to introduce many new bugs. Besides, ferries stick around for quite some time before sinking so IMHO this is not such a big problem – or just keep the ferry yard csite window open in the corner and start building waterways when you see that it is ready to start working…

stonerl commented:
@Noordfrees I analyzed  the replay myself. It wasn't the shipping algorithm. It was the users fault.... sorry for the fuzz... 😒

hessenfarmer commented:
@stonerl good to see you again here. As Klaus isn't available currently for the review maybe you could help out? 
@Noordfrees Are there any other developers we might ask to finish the review? Maybe we could get the two remaining big branches going by splitting the work for more people. I am sad to be of no help with my limited knowledge of c++.

Noordfrees commented:
Good point. The only other currently active C++ devs are @stonerl @Niektory @tppq @jmoerschbach – perhaps some of you would like continue the reviews for #3850 and/or #3857?

Noordfrees committed:
Merge branch 'master' into shipping-tweaks-b21

stonerl reviewed (commented):
I added the errors I got during compiling inline.

I'm fare away from fully understanding the code. But I'm going to continue later on, just need a break. 😄 

I did some play during the last week and I didn't find anything suspicious.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into shipping-tweaks-b21

Noordfrees committed:
Fix compiler warnings

klaus-halfmann reviewed (approved):
I finally had time to scan the rest of the code.
* I found no obvious flaws
* It was rated as an improvement
* You fixed a lot of crashes.

-> lets have it

Noordfrees commented:
Thank you for the review, and thanks everyone for the extensive testing :)

frankystone commented:
Yeah :-)
